okapi-core/src/main/java/org/folio/okapi/util/TokenCache.java

```diff
@@ -0,0 +1,181 @@
+package org.folio.okapi.util;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import org.apache.logging.log4j.Logger;
+import org.folio.okapi.common.OkapiLogger;
+
+public class TokenCache {
+
+  public static final long DEFAULT_TTL = 3 * 60 * 1000L;
+  public static final int DEFAULT_MAX_SIZE = 10000;
+
+  private final Map<String, CacheEntry> cache;
+  private final long ttl;
+
+  private static final Logger logger = OkapiLogger.get(TokenCache.class);
+
+  /**
+   * Convenience constructor using the default TTL and max size.
+   */
+  public TokenCache() {
+    this(DEFAULT_TTL, DEFAULT_MAX_SIZE);
+  }
+
+  /**
+   * Constructor using the provided TTL and maxSize.
+   * 
+   * @param ttl cache entry time to live in milliseconds
+   * @param maxSize the maximum number of entries that may be cached at once
+   */
+  private TokenCache(long ttl, int maxSize) {
+    logger.info("Initializing token cache w/ ttl: {}, maxSize: {}", ttl, maxSize);
+    this.ttl = ttl;
+
+    this.cache = Collections.synchronizedMap(new LruCache<>(maxSize));
+  }
+
+  /**
+   * Cache an entry.
+   * 
+   * @param method HTTP method
+   * @param path path pattern
+   * @param userId X-Okapi-User-Id header to cache
+   * @param xokapiPerms X-Okapi-Permissions header to cache
+   * @param keyToken the token to be used in the cache key, from the request which triggered the
+   *        call to mod-authtoken
+   * @param token access token to cache - from the mod-authtoken response
+   */
+  public void put(String tenant, String method, String path, String userId, String xokapiPerms,
+      String keyToken, String token) {
+    long now = System.currentTimeMillis();
+    CacheEntry entry = new CacheEntry(token, userId, xokapiPerms, now + ttl);
+    String key = genKey(method, path, userId, keyToken);
+    MetricsHelper.recordTokenCacheCached(tenant, method, path, userId);
+    logger.info("Caching: {} -> {}", key, token);
+    cache.put(key, entry);
+  }
+
+  /**
+   * Get a cached entry.
+   * 
+   * @param tenant tenant id
+   * @param method HTTP method
+   * @param path path pattern
+   * @param token X-Okapi-Token header
+   * @param userId X-Okapi-User-Id header
+   * @return cache entry or null
+   */
+  public CacheEntry get(String tenant, String method, String path, String userId, String token) {
+    String key = genKey(method, path, userId, token);
+    CacheEntry ret = cache.get(key);
+    if (ret == null) {
+      MetricsHelper.recordTokenCacheMiss(tenant, method, path, userId);
+      logger.info("Cache Miss: {}", key);
+      return ret;
+    } else if (ret.isExpired()) {
+      MetricsHelper.recordTokenCacheExpired(tenant, method, path, userId);
+      logger.info("Cache Hit (Expired): {}", key);
+      cache.remove(key);
+      return null;
+    } else {
+      MetricsHelper.recordTokenCacheHit(tenant, method, path, userId);
+      logger.info("Cache Hit: {} -> {}", key, ret.token);
+      return ret;
+    }
+  }
+
```

Comment:
That's going to be a lot of logging by default.. with info log-level I mean.

Reply:
OK I'll make it debug